### PR TITLE
269 followers to friends

### LINKED ISSUE #269: [269] Followers To Friends
**Current state**
Users are able to follow other users. However, when two users follow each other they still see 'Following' instead of 'Friends'.

**Expected state**
When two people follow each other it shows 'Friends' instead of 'Following'.

**TO DO**
- [ ] Task 1
- [ ] Task 2
- [ ] Task 3
- [ ] Task ...

**Necessities**
#222 

**Branch**
269-followers-to-friends

### PR DESCRIPTION
**Pull request by jerohero**
closes #269 

<br />

 **What is it supposed to do**  
The follow/unfollow button changes to follow back/unfriend if that user follows the logged in user. A _followsYou_ field was added to the return value of the GET profile route.
image

<br />

 **How can we test your branch**  
Log in with two different accounts and let them follow each other to see if the text in the follow button changes properly.

<br />

 **Additional notes**  
n/a

<br />

**1. General checklist**
- [x] a. Tested happy flow.
- [x] b. Tested unhappy flow.
- [x] c. No unexpected exceptions.
- [x] d. No code with high complexity. (Big O)
- [x] e. EOF newline.
- [x] f. No linter warnings or errors.

<br />

**2. If relevant, front-end checklist**
- [x] a. Responsive. (Desktop all the way to mobile 320px)
- [x] b. Using SCSS variables where relevant.
- [x] c. Relevant class names
- [x] d. No modified global SCSS properties, except if necessary and fully tested the impact.
- [x] e. Checked [caniuse](https://caniuse.com) for more modern CSS properties.

<br />

**3. If relevant, back-end checklist**
- [x] a. No unhandled promise rejections.
- [x] b. Seeders still functional.
- [x] c. No primary-key violations.
- [x] d. Normalization, at least [3NF](https://en.wikipedia.org/wiki/Database_normalization).
- [x] e. Used guards (middleware) where necessary and tested if route is guarded.

<br />

**4. If relevant, test these browsers**
- [x] a. Chrome
- [x] b. Firefox
- [ ] c. Opera
- [x] d. Edge
- [ ] e. Safari

### CODE REVIEW

**Comment by AjobK** [comment]
image

When a user follows me and I go to their profile it doesn't show me 'Follow Back' instead of 'Follow'

After refresh it seems to work though? Strange...

**Reproduce**
1. Register a new user
2. Follow another user
3. Logout and login on another user without refreshing
4. Go to followers by clicking on '1 followers' text and go to their profile
5. It says 'Follow' instead of 'Follow back'
6. Refresh, now it should say the correct thing

**Review by AjobK — CHANGES_REQUESTED** [review]
Check comment above

**Review by AjobK — CHANGES_REQUESTED** [review]
## Issue 1

image

When I click on follow it doesn't update it right away in the button?

Only when I refresh...

## Issue 2

image

Mobile buttons don't work well. Very unclear in which state of following you are with a person.

**Comment by jerohero** [comment]
> ## Issue 2
> image
> 
> Mobile buttons don't work well. Very unclear in which state of following you are with a person.

I didn't make this. Do you mean it doesn't represent the friend state with another user (follow back, unfriend) or the general state of following (follow, unfollow)?

**Review by ryankroon00 — DISMISSED** [review]

**Comment by jerohero** [comment]
> ## Issue 2 
> Mobile buttons don't work well. Very unclear in which state of following you are with a person.

image
image

I replaced them with some icons from Font Awesome. I don't think these icons fit very well (considering the unfollow button when the user is not followed back is a check icon, but is an icon with a slash when the user does follow back), but I was unable to find anything better.

File: frontend/src/components/userBanner/index.js [hunk 1]
```diff
@@ -106,10 +116,30 @@ class UserBanner extends Component {
...
+    const { followsYou, following } = this.state
+
+    if (followsYou) {
+      return following ? 'Unfriend' : 'Follow back'
+    } else {
+      return following ? 'Unfollow' : 'Follow'
+    }
+  }
+
+  getFollowIcon = () => {
```
Comment (AjobK) [inline frontend/src/components/userBanner/index.js]:
```suggestion
  getFollowIconName = () => {
```

File: frontend/src/components/userBanner/index.js [hunk 2]
```diff
@@ -154,8 +184,8 @@ class UserBanner extends Component {
               className={ `${ styles.follow } ${this.state.following ? styles.replied : ''}` }
               onClick={ this.follow }
             >
-              <p>{ this.state.following ? 'unfollow' : 'follow' }</p>
-              <Icon iconName={ this.state.following ? 'Check' : 'Reply' } classNames={ styles.replyIcon } />
+              <p>{ this.getFollowText() }</p>
+              <Icon iconName={ this.getFollowIcon() } classNames={ styles.replyIcon } />
```
Comment (AjobK) [inline frontend/src/components/userBanner/index.js]:
```suggestion
              <Icon iconName={ this.getFollowIconName() } classNames={ styles.replyIcon } />
```

File: frontend/src/components/userBanner/index.js [hunk 3]
```diff
@@ -106,10 +116,30 @@ class UserBanner extends Component {
       })
   }
 
-  banUser() {
+  banUser = () => {
     this.setState({ banUser: true })
   }
```
Comment (AjobK) [inline frontend/src/components/userBanner/index.js:123]: banUser is literally nothing more than a state change?

File: frontend/src/components/userBanner/index.js [hunk 4]
```diff
@@ -17,11 +17,21 @@ class UserBanner extends Component {
...
     }
   }
 
+  componentDidUpdate(prevProps) {
+    if (this.props.user.username !== prevProps.user.username) {
+      this.setState({
+        following: this.props.user.following,
+        followsYou: this.props.user.followsYou
+      })
+    }
```
Comment (AjobK) [inline frontend/src/components/userBanner/index.js]:
```suggestion
    const { username, following, followsYou } = this.props
    
    if (username !== username) {
      this.setState({
        following,
        followsYou
      })
    }
```

Deconstruct and if I'm correct, state doesn't have to specify var name if variable is same as state property name

File: backend/app/controllers/profileController.ts [hunk 5]
```diff
@@ -167,19 +167,28 @@ class ProfileController {
     if (receivedUsername && decodedToken) isOwner = receivedUsername == decodedToken.username
 
     let following = false
+    let followsYou = false
 
     if (!isOwner && decodedToken && decodedToken.username) {
-      const profileFollowedBy: ProfileFollowedBy = new ProfileFollowedBy()
+      const visitingProfile: ProfileFollowedBy = new ProfileFollowedBy()
+      const ownProfile: ProfileFollowedBy = new ProfileFollowedBy()
```
Comment (AjobK) [inline backend/app/controllers/profileController.ts]: I don't understand this var name..? Maybe something more specific?

**Review by AjobK — CHANGES_REQUESTED** [review]

File: frontend/src/components/userBanner/index.js (hunk 3 above)
Comment (jerohero) [reply frontend/src/components/userBanner/index.js:123]: Didn't make this, just felt cute and changed the function into an arrow function. It's replaced now.

File: backend/app/controllers/profileController.ts (hunk 5 above)
Comment (jerohero) [reply backend/app/controllers/profileController.ts]: It's the profile the user is visiting and the own profile of the user. I changed _visitingprofile_ into _targetProfile_. Maybe that's more specific?

**Review by ryankroon00 — DISMISSED** [review]

**Review by ryankroon00 — APPROVED** [review]

**Review by AjobK — APPROVED** [review]

**Pull request merged**